[WIP] enhancement: WIP add new collaboration service (WOPI) (#8374)
owncloud/ocis: services/collaboration/README.md

# Collaboration

The collaboration service connects ocis with document servers such as collabora and onlyoffice using the WOPI protocol.

> Since this service requires an external service (onlyoffice, for example), it won't run by default with the general `ocis server` command. You need to run it manually with the `ocis collaboration server` command.

Author: Yes. I'll include the info in the README.

## Requirements

The collaboration service requires the target document server (onlyoffice, collabora, etc) to be up and running.
We also need reva's gateway and app provider services to be running in order to register the GRPC service for the "open in app" action.

If any of those services are down, the collaboration service won't start.

## Configuration

There are a few variables that you need to set:

* `COLLABORATION_WOPIAPP_ADDR`: The URL of the WOPI app (onlyoffice, collabora, etc). For example: "https://office.mycloud.prv".
* `COLLABORATION_HTTP_ADDR`: The external address of the collaboration service. The target app (onlyoffice, collabora) will use this address to read and write files from ocis. For example: "wopiserver.mycloud.prv"

> * `COLLABORATION_HTTP_SCHEME`: The scheme to be used when accessing the collaboration service. Either "http" or "https". This will be used to build the URL that the WOPI app needs in order to contact this service.

Reviewer: ```suggestion
* `COLLABORATION_HTTP_SCHEME`: The scheme to be used when accessing the collaboration service. Either "http" or "https". This will be used to finally build the URL that the WOPI app needs in order to contact this service.
```

> The rest of the configuration options available can be left with the default values.

Reviewer 2: I couldn't run `ocis collaboration` following readmi.

> The collaboration service requires the target document server (onlyoffice, collabora, etc) to be up and running

I used `onlyoffice` which I started here https://github.com/owncloud/ocis/blob/master/deployments/examples/ocis_wopi/docker-compose.yml

I started `ocis server` and  `collaboration server` with envs `COLLABORATION_WOPIAPP_ADDR=https://onlyoffice.owncloud.test COLLABORATION_HTTP_ADDR=onlyoffice.owncloud.test COLLABORATION_HTTP_SCHEME=https`

and get: 

```
2024-04-09T11:10:27+02:00 ERR WopiDiscovery: failed to access wopi app url error="Get \"https://onlyoffice.owncloud.test/hosting/discovery\": tls: failed to verify certificate: x509: certificate signed by unknown authority" WopiAppUrl=https://onlyoffice.owncloud.test/hosting/discovery service=collaboration
Get "https://onlyoffice.owncloud.test/hosting/discovery": tls: failed to verify certificate: x509: certificate signed by unknown authority

```

could you plz say, what I did wrong?

Author: ```
COLLABORATION_HTTP_ADDR=wopiserver:9300
COLLABORATION_HTTP_SCHEME=http
COLLABORATION_WOPIAPP_ADDR=https://onlyoffice.owncloud.test
```
The `HTTP_*` refers to this collaboration service. The `WOPI_*` refers to onlyoffice (or other WOPI app).
I think the `HTTP_ADDR` doesn't need to be fully external (like `https//wopiserver.owncloud.test`) but it needs to be accessible to the WOPI app (onlyoffice).

I haven't tested with certificates. There are a couple of `_INSECURE` options to disable certificate validation, so maybe you can try without certificates first to check that you can connect without problems.
The error shows a problem verifying the certificate of the onlyoffice server.

Reviewer 2: > ```
> COLLABORATION_HTTP_ADDR=wopiserver:9300
> COLLABORATION_HTTP_SCHEME=http
> COLLABORATION_WOPIAPP_ADDR=https://onlyoffice.owncloud.test
> ```

after many trying I couldn't run `ocis collaboration` following readmi, sorry seems too complicated to me. 
but this way https://github.com/owncloud/ocis/pull/8374#issuecomment-2044726955 works fine to me

# Collaboration

The collaboration service connects ocis with document servers such as collabora and onlyoffice using the WOPI protocol.

Since this service requires an external service (onlyoffice, for example), it won't run by default with the general `ocis server` command. You need to run it manually with the `ocis collaboration server` command.

## Requirements

The collaboration service requires the target document server (onlyoffice, collabora, etc) to be up and running.

> Some oCIS services are also required to be running in order to register the GRPC service for the "open in app" action:
> * reva's gateway
> * app provider

Reviewer: ```suggestion
The following oCIS services are required to be running for the collaboration service in order to register the GRPC service for the `open in app` action:
* gateway
* app-provider
```
There is no `reva's gateway`but a gateway service. Also correcting the name of the app-provider service.

Author: It refers to the gateway API client in https://github.com/owncloud/ocis/pull/8374/files#diff-f02696d0615274e7591a84f42effbb7ab340a9b5c66976fe7f8a5ee8b5c80d92 . Since the code comes from the cs3org, I assumed we'd be connecting to a reva's gateway, but I could be wrong. Maybe @butonic can clarify.

If any of those services are down, the collaboration service won't start.

> For convenience, you might want to start the whole oCIS stack before starting the collaboration service.

Reviewer: This is an interresting sentence as it highlights the possibility that the collaboration service is NOT part of the services started up automatically in the binary build. I may be wrong...

Author: I think it's commented above, in the second paragraph of the readme. We'll need to connect to an external server (onlyoffice, collabora, etc) which might not be available, and it will require configuration (we can't provide a valid default), that's why you need to start the service manually.
